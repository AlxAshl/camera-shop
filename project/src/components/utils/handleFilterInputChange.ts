import { ActionCreatorWithNonInferrablePayload } from '@reduxjs/toolkit';
import { store } from '../../store/store';

export const handleFilterInputChange = (filterName: string, filter: string[], action: ActionCreatorWithNonInferrablePayload) => {
  const mutableFilter = Object.assign([], filter);
  if(filter.includes(filterName)) {
    const updatedFilter = mutableFilter.filter((entry) => entry !== filterName);
    store.dispatch(action(updatedFilter));
  }
  else {
    mutableFilter.push(filterName);
    store.dispatch(action(mutableFilter));
  }
};
