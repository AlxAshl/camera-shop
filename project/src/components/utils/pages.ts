import { PAGE_LIMIT } from '../../const';

function getPagesCount(cameraCount: number): number {
  return Math.ceil(cameraCount / PAGE_LIMIT);
}

export default getPagesCount;
