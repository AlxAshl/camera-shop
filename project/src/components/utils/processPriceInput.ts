import { ActionCreatorWithNonInferrablePayload } from '@reduxjs/toolkit';
import { store } from '../../store/store';
import { ProductType } from '../../types/product';

export const setMinInput = (input: string, productsArray: ProductType[], action: ActionCreatorWithNonInferrablePayload) => {
  if(productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    store.dispatch(action([String(input)]));
  }
  if(!productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    let newMinPriceInput;
    const newMin = productsArray.find((product) => product.price <= Number(input));
    newMin?.price === undefined
      ? newMinPriceInput = productsArray[productsArray.length - 1].price
      : newMinPriceInput = newMin?.price;
    store.dispatch(action([String(newMinPriceInput)]));
  }
};

export const setMaxInput = (input: string, productsArray: ProductType[], action: ActionCreatorWithNonInferrablePayload) => {
  if(productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    store.dispatch(action([String(input)]));
  }
  if(!productsArray.find((product) => product.price === Number(input)) && productsArray.length !== 0) {
    let newMaxPriceInput;
    const newMax = productsArray.find((product) => product.price >= Number(input));
    newMax?.price === undefined
      ? newMaxPriceInput = productsArray[productsArray.length - 1].price
      : newMaxPriceInput = newMax?.price;
    store.dispatch(action([String(newMaxPriceInput)]));
  }
};
