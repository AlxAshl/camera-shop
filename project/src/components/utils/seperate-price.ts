export function seperatePrice(priceValue: number) {
  let returnedPrice = String(priceValue);
  if(returnedPrice.length > 6) {
    returnedPrice = `${returnedPrice.slice(0, 1)} ${returnedPrice.slice(1, 4)} ${returnedPrice.slice(-3)}`;
  }
  if (returnedPrice.length > 3) {
    returnedPrice = `${returnedPrice.slice(-0, -3)} ${returnedPrice.slice(-3)}`;
  }
  return returnedPrice;
}
