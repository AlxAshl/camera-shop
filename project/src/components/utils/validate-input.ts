export function validateInput(target: (EventTarget & HTMLInputElement) | (EventTarget & HTMLTextAreaElement)) {

  if (target.parentElement?.parentElement && target.name === 'rating' && (Number(target.value) > 0)) {
    if (!target.parentElement.parentElement.parentElement?.className.includes('is-valid')) {
      target.parentElement.parentElement.parentElement?.classList.add('is-valid');
      target.parentElement.parentElement.parentElement?.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && target.name !== 'review' && typeof(target.value) === 'string') {
    if (!target.parentElement.parentElement.className.includes('is-valid') && target.value.length > 0) {
      target.parentElement.parentElement.className += ' is-valid';
      target.parentElement.parentElement.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && typeof(target.value) === 'string'){
    if(target.value.length === 0) {
      target.parentElement.parentElement.className += ' is-invalid';
      target.parentElement.parentElement.classList.remove('is-valid');
    }
  }

  if (target.parentElement?.parentElement && typeof(target.value) === 'string') {
    if (!target.parentElement.parentElement.className.includes('is-valid') && target.value.length >= 5) {
      target.parentElement.parentElement.className += ' is-valid';
      target.parentElement.parentElement.classList.remove('is-invalid');
    }
  }

  if (target.parentElement?.parentElement && target.name === 'review' && target.value.length < 5) {
    if (!target.parentElement.parentElement.className.includes('is-invalid')) {
      target.parentElement.parentElement.className += ' is-invalid';
      target.parentElement.parentElement.classList.remove('is-valid');
    }
  }

}
