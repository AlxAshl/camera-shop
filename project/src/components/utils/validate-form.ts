import { reviewDataType } from '../../types/review';
import { validateInput } from './validate-input';


export function validateForm(data: reviewDataType) {
  let validationStatus = true;
  Object.entries(data).forEach(([key, value]) =>{
    if (key !== 'review') {
      if(value === '') {
        const failedInput = document.querySelector(`input[name=${key}]`);
        validateInput(failedInput as HTMLInputElement);
        validationStatus = false;
      }
    }
    if(key === 'review') {
      if(String(value).length < 5) {
        const failedInput = document.querySelector(`textarea[name=${key}]`);
        validateInput(failedInput as HTMLInputElement);
      }
    }
    if(key === 'rating') {
      if(value < 1) {
        const failedInput = document.querySelector(`input[name=${key}]`);
        if(failedInput && failedInput.parentElement?.parentElement){
          failedInput.parentElement.parentElement.parentElement?.classList.add('is-invalid');
        }
        validationStatus = false;
      }
    }
  });
  return validationStatus;
}

