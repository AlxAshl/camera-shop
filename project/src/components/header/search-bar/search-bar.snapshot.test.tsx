
import { renderWithProviders } from '../../../test/utils/render-with-redux';
import SearchBar from './search-bar';

test('Render SearchBar', () => {
  const { asFragment } = renderWithProviders(<SearchBar />);
  expect(asFragment()).toMatchSnapshot();
});
