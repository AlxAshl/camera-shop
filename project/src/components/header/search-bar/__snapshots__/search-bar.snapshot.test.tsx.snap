// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render SearchBar 1`] = `
<DocumentFragment>
  <div
    class="form-search "
    data-testid="search-form-test"
    id="search-form"
  >
    <form>
      <label>
        <svg
          aria-hidden="true"
          class="form-search__icon"
          height="16"
          width="16"
        >
          <use
            xlink:href="#icon-lens"
          />
        </svg>
        <input
          autocomplete="off"
          class="form-search__input"
          data-testid="search-input-test"
          id="search"
          placeholder="Поиск по сайту"
          type="text"
          value=""
        />
      </label>
    </form>
    <button
      class="form-search__reset"
      data-testid="search-reset-test"
      id="reset"
      type="reset"
    >
      <svg
        aria-hidden="true"
        height="10"
        width="10"
      >
        <use
          xlink:href="#icon-close"
        />
      </svg>
      <span
        class="visually-hidden"
      >
        Сбросить поиск
      </span>
    </button>
  </div>
</DocumentFragment>
`;
